src/System.Private.CoreLib/shared/System/Buffers/Text/Utf8Parser/Utf8Parser.Date.I.cs

```diff
@@ -0,0 +1,305 @@
+                goto ReturnFalse;
+            }
+
+            // We now have YYYY-MM-DDThh:mm
+            bytesConsumed = 16;
+
+            if (source.Length < 17)
+            {
+                goto FinishedParsing;
+            }
+
+            curByte = source[16];
+
+            int sourceIndex = 16;
+
+            if (curByte == Utf8Constants.UtcOffsetToken)
+            {
+                bytesConsumed++;
+                offsetToken = Utf8Constants.UtcOffsetToken;
+                goto FinishedParsing;
+            }
+            else if (curByte == Utf8Constants.Plus || curByte == Utf8Constants.Minus)
+            {
+                offsetToken = curByte;
+                sourceIndex++;
+                goto ParseOffset;
+            }
+            else if (curByte != Utf8Constants.Colon)
+            {
+                goto FinishedParsing;
+            }
+
+            if (!ParserHelpers.TryGetNextTwoDigits(source.Slice(start: 17, length: 2), out second))
+            {
+                goto ReturnFalse;
+            }
+
+            // We now have YYYY-MM-DDThh:mm:ss
+            bytesConsumed = 19;
+
+            if (source.Length < 20)
+            {
+                goto FinishedParsing;
+            }
+
+            curByte = source[19];
+            sourceIndex = 19;
+
+            if (curByte == Utf8Constants.UtcOffsetToken)
+            {
+                bytesConsumed++;
+                offsetToken = Utf8Constants.UtcOffsetToken;
+                goto FinishedParsing;
+            }
+            else if (curByte == Utf8Constants.Plus || curByte == Utf8Constants.Minus)
+            {
+                offsetToken = curByte;
+                sourceIndex += 1;
+                goto ParseOffset;
+            }
+            else if (curByte != Utf8Constants.Period)
+            {
+                goto FinishedParsing;
+            }
+
+            // Source does not have enough characters for YYYY-MM-DDThh:mm:ss.s
+            if (source.Length < 21)
+            {
+                value = default;
+                bytesConsumed = 0;
+                kind = default;
+                return false;
+            }
+
+            sourceIndex = 20;
+
+            {
+                while (sourceIndex < source.Length && ParserHelpers.IsDigit(curByte = source[sourceIndex]))
+                {
+                    int prevFractionTimesTen = fraction * 10;
+
+                    if (!(prevFractionTimesTen + (int)(curByte - (uint)'0') <= Utf8Constants.MaxDateTimeFraction))
+                    {
+                        sourceIndex++;
+                        break;
+                    }
+
+                    fraction = prevFractionTimesTen + (int)(curByte - (uint)'0');
+                    sourceIndex++;
+                }
+            }
+
+            if (fraction != 0)
+            {
+                // Note this is 1 order of magnitude less than Utf8Constants.MaxDateTimeFraction
+                while (fraction <= 999_999)
```

Comment:
nit: use named constant for this. Maybe `MaxDateTimeFractionDiv10` or something.